Neos.Diff/Classes/Renderer/Html/HtmlArrayRenderer.php

```diff
@@ -0,0 +1,200 @@
+                            'lines' => array()
+                        ),
+                        'changed' => array(
+                            'offset' => $j1,
+                            'lines' => array()
+                        )
+                    );
+                    $lastBlock = count($blocks) - 1;
+                }
+
+                $lastTag = $tag;
+
+                if ($tag == 'equal') {
+                    $lines = array_slice($a, $i1, ($i2 - $i1));
+                    $blocks[$lastBlock]['base']['lines'] += $this->formatLines($lines);
+                    $lines = array_slice($b, $j1, ($j2 - $j1));
+                    $blocks[$lastBlock]['changed']['lines'] += $this->formatLines($lines);
+                } else {
+                    if ($tag == 'replace' || $tag == 'delete') {
+                        $lines = array_slice($a, $i1, ($i2 - $i1));
+                        $lines = $this->formatLines($lines);
+                        $lines = str_replace(array("\0", "\1"), array('<del>', '</del>'), $lines);
+                        $blocks[$lastBlock]['base']['lines'] += $lines;
+                    }
+
+                    if ($tag == 'replace' || $tag == 'insert') {
+                        $lines = array_slice($b, $j1, ($j2 - $j1));
+                        $lines = $this->formatLines($lines);
+                        $lines = str_replace(array("\0", "\1"), array('<ins>', '</ins>'), $lines);
+                        $blocks[$lastBlock]['changed']['lines'] += $lines;
+                    }
+                }
+            }
+            $changes[] = $blocks;
+        }
+        return $changes;
+    }
+
+    /**
+     * Given two strings, determine where the changes in the two strings
+     * begin, and where the changes in the two strings end.
+     *
+     * @param string $fromLine The first string.
+     * @param string $toLine The second string.
+     * @return array Array containing the starting position (0 by default) and the ending position (-1 by default)
+     */
+    private function getChangeExtent($fromLine, $toLine)
+    {
+        $start = 0;
+        $limit = min(strlen($fromLine), strlen($toLine));
+        while ($start < $limit && $fromLine{$start} == $toLine{$start}) {
+            ++$start;
+        }
+        $end = -1;
+        $limit = $limit - $start;
+        while (-$end <= $limit && substr($fromLine, $end, 1) == substr($toLine, $end, 1)) {
+            --$end;
+        }
+        return array(
+            $start,
+            $end + 1
+        );
+    }
+
+    /**
+     * Format a series of lines suitable for output in a HTML rendered diff.
+     * This involves replacing tab characters with spaces, making the HTML safe
+     * for output, ensuring that double spaces are replaced with &nbsp; etc.
+     *
+     * @param array $lines Array of lines to format.
+     * @return array Array of the formatted lines.
+     */
+    private function formatLines($lines)
```

Comment:
If you like ...

On one occasion I tried to add an array type hint elsewhere in the code and it turned out that it *sometimes* also was a string rather than an array. Well.